[git-webkit] Use graphQL API to search for pull requests

JonWBedard opened the pull request:
#### 500b3dbc7c61814787e117e9e724f990f9fa2ec2
<pre>
[git-webkit] Use graphQL API to search for pull requests
<a href="https://bugs.webkit.org/show_bug.cgi?id=242927">https://bugs.webkit.org/show_bug.cgi?id=242927</a>
&lt;rdar://problem/97290312&gt;

Reviewed by Aakash Jain and Ryan Haddad.

* Tools/Scripts/libraries/webkitscmpy/setup.py: Bump version.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/__init__.py: Ditto.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/mocks/remote/git_hub.py:
(GitHub.graphql): Added.
(GitHub.request): Add graphql endpoint.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/remote/git_hub.py:
(GitHub.PRGenerator.find): Use GraphQL API instead of REST API. The GraphQL API is faster and
allows for more flexible queries.
(GitHub.graphql): Added.

Canonical link: <a href="https://commits.webkit.org/252661@main">https://commits.webkit.org/252661@main</a>
</pre>

youssefsoli commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/mocks/remote/git_hub.py:
```diff
@@ -289,9 +290,69 @@ def _parents_of_request(self, url, ref):
-        from datetime import datetime, timedelta
+    # FIXME: Not a very flexible mock of GitHub's GraphQL API, only supports pull-request querying
+    def graphql(self, url, auth=None, json=None):
+        query = (json or {}).get('query')
+        if not query:
+            return mocks.Response.create404(url)
```
```py
if not json:
   return mocks.Response.create404(url)
query = json.get('query')
```
might look a bit cleaner

youssefsoli replied:
Never mind, this covers the case if `query` is an empty string.

youssefsoli commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/mocks/remote/git_hub.py:
```diff
@@ -289,9 +290,69 @@ def _parents_of_request(self, url, ref):
+        if not query:
+            return mocks.Response.create404(url)
 
+        qline = query.splitlines()[1]
```
It might make more sense to parse the query as an object instead of splitting it based off newlines since a valid GraphQL query doesn't necessarily have to contain any newlines, it will also reduce the error rate and responsibilities of the caller having a "properly" formed query.

Parsing it as an object will also make this mock much more extensible for other queries if we decide to move other REST calls to GraphQL.

JonWBedard replied:
This is a unit test, so we're only making an effort to make sense of a very specific graphQL query. I'm desperately trying to avoid implementing an actual graphQL parser. All we really need in this case is something that can respond to the production query we're actually sending GitHub.

youssefsoli replied:
Fair, I am just worried about future contributors coming across this method and having difficulties capturing and parsing their request properly, maybe leave a TODO for future contributors?

JonWBedard replied:
There is the `FIXME` right above the function declaration.

youssefsoli commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/mocks/remote/git_hub.py:
```diff
@@ -289,9 +290,69 @@ def _parents_of_request(self, url, ref):
+                elif bit == 'is:open':
+                    state = 'open'
+                elif bit == 'is:closed':
+                    state = 'closed'
```
Maybe move this query parsing to a separate function?

facetothefate commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/remote/git_hub.py:
```diff
@@ -359,6 +381,21 @@ def request(self, path=None, params=None, headers=None, authenticated=None, pagi
+            sys.stderr.write("Request to '{}' returned status code '{}'\n".format(url, response.status_code))
+            message = response.json().get('message')
+            if message:
+                sys.stderr.write('Message: {}\n'.format(message))
```
This message maybe hard to understand for user to know what should he do to fix it, maybe for any 401 we suggest user to rerun setup ?

JonWBedard replied:
That's what the `REFRESH_TOKEN_PROMPT` on the line right bellow this one will do.

The reason I think we should continue to print errors here is that if folks persistently encounter such errors, if we don't print them, we have no way of knowing why a particular tools invocation is failing.

youssefsoli commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/mocks/remote/git_hub.py:
```diff
@@ -451,4 +512,8 @@ def request(self, method, url, data=None, params=None, auth=None, json=None, **k
             return self.releases.get(stripped_url[len(download_base):], mocks.Response.create404(url))
 
+        # GraphQL library
+        if method == 'POST' and auth and stripped_url == '{}/graphql'.format(self.api_remote.split('/')[0]):
```
Is it possible for `api_remote` to not be GitHub? If so, this will fail for remote API's that don't support GraphQL.

JonWBedard replied:
This is a mock implementation of a limited set of GitHub's API. So we can be sure that the `api_remote` is GitHub.

youssefsoli commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/remote/git_hub.py:
```diff
@@ -74,32 +79,49 @@ def get(self, number):
+            }.get(opened, '')
+
+            data = self.repository.graphql('''query {{
+  search(query: "{}", type: ISSUE, last: 100) {{
```
We no longer paginate if we have more than 100 potential PRs right?

JonWBedard replied:
Right, the reason I originally did the pagination thing is because GitHub doesn't let you ask for a PR by `head` name unless the `head` name includes the user. For some reason, this API doesn't work reliably in private repositories (meaning we can't find an existing PR), but the graphQL API does work.

youssefsoli replied:
In that case, should we expect the API to always return one match? What is the 100 limit based off? (other than that being the API maximum per query)

JonWBedard replied:
We can expect the API to return some matches, mostly likely 0 or 1, but if a branch name has been repeated, possibly more. Unlikely to be more than a dozen. If we need to support more interesting queries going forward, I'd sort of like for this testing function to be made a bit more robust. For now, the goal of this PR is to support a relatively specific use case involving a private repository.

youssefsoli commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/remote/git_hub.py:
```diff
@@ -74,32 +79,49 @@ def get(self, number):
+            search += {
+                True: ' is:open',
+                False: ' is:closed',
+            }.get(opened, '')
```
```py
if opened is not None:
    search += ' is:open' if opened else ' is:closed'
```

youssefsoli commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/mocks/remote/git_hub.py:
```diff
@@ -289,9 +290,66 @@ def _parents_of_request(self, url, ref):
+
+        qline = query.splitlines()[1]
+        pr_search = re.match(r'\s*search\(query:\s+"(?P<query>.+)",\s+type:\s+ISSUE,\s+last:\s+(?P<last>\d+)\)\s*\{', qline)
+        if pr_search:
```
Maybe
```py
if not pr_search:
    return mocks.Response.create404(url)
```
to reduce the amount of indentation

JonWBedard replied:
I want to keep it like this because if folks want to add more queries to this function (which, regrettably, seems quite likely), they'll just end up re-adding the indentation.

JonWBedard committed:
[git-webkit] Use graphQL API to search for pull requests
https://bugs.webkit.org/show_bug.cgi?id=242927
<rdar://problem/97290312>

Reviewed by Aakash Jain and Ryan Haddad.

* Tools/Scripts/libraries/webkitscmpy/setup.py: Bump version.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/__init__.py: Ditto.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/mocks/remote/git_hub.py:
(GitHub.graphql): Added.
(GitHub.request): Add graphql endpoint.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/remote/git_hub.py:
(GitHub.PRGenerator.find): Use GraphQL API instead of REST API. The GraphQL API is faster and
allows for more flexible queries.
(GitHub.graphql): Added.

Canonical link: https://commits.webkit.org/252661@main
